Source/Core/Core/PowerPC/JitArm64/JitAsm.cpp

```diff
@@ -272,6 +281,84 @@ void JitArm64::GenerateConvertSingleToDouble()
   RET();
 }
 
+// Input in X0. Outputs to memory (PPCState). Clobbers X0-X4 and flags.
+void JitArm64::GenerateFPRF(bool single)
+{
+  const auto reg_encoder = single ? EncodeRegTo32 : EncodeRegTo64;
+
+  const ARM64Reg input_reg = reg_encoder(ARM64Reg::W0);
+  const ARM64Reg temp_reg = reg_encoder(ARM64Reg::W1);
+  const ARM64Reg exp_reg = reg_encoder(ARM64Reg::W2);
+  const ARM64Reg frac_reg = reg_encoder(ARM64Reg::W3);
+
+  const auto INPUT_EXP_MASK = single ? Common::FLOAT_EXP : Common::DOUBLE_EXP;
+  const auto INPUT_FRAC_MASK = single ? Common::FLOAT_FRAC : Common::DOUBLE_FRAC;
+  constexpr u32 OUTPUT_SIGN_MASK = 0xC;
+
+  // First of all, start the load of the old FPSCR value, in case it takes a while
+  LDR(IndexType::Unsigned, ARM64Reg::W4, PPC_REG, PPCSTATE_OFF(fpscr));
+
+  CMP(input_reg, 0);  // Grab sign bit (conveniently the same bit for floats as for integers)
+  ANDI2R(exp_reg, input_reg, INPUT_EXP_MASK);    // Grab exponent
+  ANDI2R(frac_reg, input_reg, INPUT_FRAC_MASK);  // Grab mantissa
+
+  // Most branches handle the sign in the same way. Perform that handling before branching
+  MOVI2R(ARM64Reg::W0, Common::PPC_FPCLASS_PN);
+  MOVI2R(ARM64Reg::W1, Common::PPC_FPCLASS_NN);
+  CSEL(ARM64Reg::W0, ARM64Reg::W1, ARM64Reg::W0, CCFlags::CC_LT);
+
+  FixupBranch zero_exponent = CBZ(exp_reg);
+
+  MOVI2R(temp_reg, INPUT_EXP_MASK);
+  CMP(exp_reg, temp_reg);
+  FixupBranch normal = B(CCFlags::CC_NEQ);
```

Comment:
So, I was actually thinking about that earlier. At first impression I assumed that it would save one instruction while not being slower, but Clang doesn't seem to want to output that kind of sequence (I've used Clang's output as a reference for this code), and checking whether two numbers are equal feels like such a common thing to do that I think Clang would've made this optimization by now if this actually was better... I'm not sure why the current code would actually be better, though. Are there maybe some CPUs out there that fuse CMP with a following conditional branch, giving the current code lower latency than the EOR alternative?

Reply:
Yeah, it just saves one instruction, not a u-op. CBZ is actually more expensive than B. So as long as the immediate generation is for free, the CBZ approach might actually be slower.

Reply:
I think I'll keep it as it is then. Instruction count isn't a big concern, since this code is only emitted twice ever.